[LoopInterchange] Update the direction of undistributed loop to EQ

ShivaChen opened the pull request:
This commit enables loop-interchange for the case in https://github.com/llvm/llvm-project/issues/71519.
With the loop-interchange, the case can be vectorized.

      for (int nl = 0; nl < 10000000/256; nl++) // Level 1
        for (int i = 0; i < 256; ++i)           // Level 2
          for (int j = 1; j < 256; j++)         // Level 3
            aa[j][i] = aa[j - 1][i] + bb[j][i];

The case can't be interchanged without normalizaion.
normalizaion didn't occur because the direction of level 1 loop dependence between aa[j][i] and aa[j - 1][i] is default value '*'.

By scanning SCEV form of the pointer of aa[j][i] and aa[j - 1][i], the pass can determine the IV of loop 1(nl) didn't affect the value of aa[j][i] and aa[j - 1][i]. And then updating the direction of loop 1 to '=' to enable the normalization.

ShivaChen committed:
Add pr71519.ll

Meinersbur commented:
I would need a more convincing argument than "it makes my case work".

Suppose the scalar dependency is an implicit `<` dependencies. 
```
for (int i = 0; i < n; ++i)
  a=a/i;
```
This example would not be reversed because there is no inner level but consider this to be the "either way goes" case.
Reversing the scalar dependence would make a `>`-dependence out of it, which is definitely wrong because it would make it depend on the future (which `normalize()` is supposed to normalize).

Possibly, `normalize()` would also need to reverse the direction of scalar dependencies, but I would need some more thoughts on why

bmahjour reviewed (changes requested):
Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence. 

I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.

ShivaChen commented:
> Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence.
> 
> I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.

Thanks for pointing the direction. :-)  Could you help me to check does the latest commit as you expected?

bmahjour commented:
> > Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence.
> > I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.
> 
> Thanks for pointing the direction. :-) Could you help me to check does the latest commit as you expected?

I'm saying let's not modify DependenceAnalysis. The custom handling should be done in Loop Interchange to ignore dependencies at outer levels when they don't matter to the legality of interchange.

ShivaChen commented:
> > > Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence.
> > > I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.
> > 
> > 
> > Thanks for pointing the direction. :-) Could you help me to check does the latest commit as you expected?
> 
> I'm saying let's not modify DependenceAnalysis. The custom handling should be done in Loop Interchange to ignore dependencies at outer levels when they don't matter to the legality of interchange.

I try to implement without modifying DependenceAnalysis and figure out that LoopInterchange rely on normalize() to enable few optimization opportunities. normalize() didn't happen because the outer loop is the initial value DVEntry::ALL.
Currently, normalize() is part of DependenceAnalysis which allow the function to update private DV variable.

Setting non-constributed(IV didn't affect instructions) loops direction to NONE for each dependency by scanning the SCEV form of the pointers of Src and Dst instructions instead of initial value ALL seems generally correct for DependenceAnalysis.

Or LoopInterchange should explore the way to legalize interchange without normalize? Thanks.

bmahjour commented:
> > > > Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence.
> > > > I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.
> > > 
> > > 
> > > Thanks for pointing the direction. :-) Could you help me to check does the latest commit as you expected?
> > 
> > 
> > I'm saying let's not modify DependenceAnalysis. The custom handling should be done in Loop Interchange to ignore dependencies at outer levels when they don't matter to the legality of interchange.
> 
> I try to implement without modifying DependenceAnalysis and figure out that LoopInterchange rely on normalize() to enable few optimization opportunities. normalize() didn't happen because the outer loop is the initial value DVEntry::ALL. Currently, normalize() is part of DependenceAnalysis which allow the function to update private DV variable.
> 
> Setting non-constributed(IV didn't affect instructions) loops direction to NONE for each dependency by scanning the SCEV form of the pointers of Src and Dst instructions instead of initial value ALL seems generally correct for DependenceAnalysis.
> 
> Or LoopInterchange should explore the way to legalize interchange without normalize? Thanks.

What would the semantics of `DVEntry::NONE` be? It cannot be treated less pessimistically than `ALL` because the same memory locations get modified in every iteration of the outer loop! It's important for DenepdenceAnalysis to remain conservatively correct in the results in produces, regardless of the consumer of that analysis. 

IMHO, how loop interchange deals with the "non-contributing" outer loops is an implementation issue inside loop-interchange itself. One solution might be to have its own normalize function, so you construct the dependence matrix, prune it (eg. by replacing the `*` with `=` in that entry), then normalize it again.

ShivaChen committed:
[LoopInterchange] Add normalize function from DependenceInfo

The motivation is to introduce the custom functions for LoopInterchange.

ShivaChen committed:
[LoopInterchange] Update the direction of undistributed loop to EQ

This commit enables loop-interchange for the case in #71519.
With the loop-interchange, the case can be vectorized.

  for (int nl = 0; nl < 10000000/256; nl++) // Level 1
    for (int i = 0; i < 256; ++i)           // Level 2
      for (int j = 1; j < 256; j++)         // Level 3
        aa[j][i] = aa[j - 1][i] + bb[j][i];

The case can't be interchanged without normalizaion.
normalizaion didn't occur because the direction of level 1 loop dependence
between aa[j][i] and aa[j - 1][i] is default value '*'.

By scanning SCEV form of the pointer of aa[j][i] and aa[j - 1][i], the pass
and determine the IV of loop 1(nl) didn't affect the value of aa[j][i] and
aa[j - 1][i]. And then updating the direction of loop 1 to '=' to enable
the normalization.

ShivaChen commented:
> > > > > Skipping the scalar direction is incorrect. Scalar dependencies get carried from every iteration to every iteration, so they **must** be treated more conservatively than a `=` dependence.
> > > > > I think this problem should be solved in the loop interchange pass itself where if the candidate pair of loops is nested inside other loops, and those outer loop IVs don't contribute to the index calculations in the pair of loops, then the dependence entries corresponding to the outer loops should be ignored.
> > > > 
> > > > 
> > > > Thanks for pointing the direction. :-) Could you help me to check does the latest commit as you expected?
> > > 
> > > 
> > > I'm saying let's not modify DependenceAnalysis. The custom handling should be done in Loop Interchange to ignore dependencies at outer levels when they don't matter to the legality of interchange.
> > 
> > 
> > I try to implement without modifying DependenceAnalysis and figure out that LoopInterchange rely on normalize() to enable few optimization opportunities. normalize() didn't happen because the outer loop is the initial value DVEntry::ALL. Currently, normalize() is part of DependenceAnalysis which allow the function to update private DV variable.
> > Setting non-constributed(IV didn't affect instructions) loops direction to NONE for each dependency by scanning the SCEV form of the pointers of Src and Dst instructions instead of initial value ALL seems generally correct for DependenceAnalysis.
> > Or LoopInterchange should explore the way to legalize interchange without normalize? Thanks.
> 
> What would the semantics of `DVEntry::NONE` be? It cannot be treated less pessimistically than `ALL` because the same memory locations get modified in every iteration of the outer loop! It's important for DenepdenceAnalysis to remain conservatively correct in the results in produces, regardless of the consumer of that analysis.
> 
> IMHO, how loop interchange deals with the "non-contributing" outer loops is an implementation issue inside loop-interchange itself. One solution might be to have its own normalize function, so you construct the dependence matrix, prune it (eg. by replacing the `*` with `=` in that entry), then normalize it again.

Have created normalize function in interchange. Thanks for the advice. :-)

ShivaChen committed:
Fix clang-format

ShivaChen commented:
Fix clang-format and rebase.

ShivaChen committed:
Add missing LLVM_DEBUG

ShivaChen committed:
Replace Levels parameter by DV.size()

madhur13490 commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -82,6 +82,101 @@ static void printDepMatrix(CharMatrix &DepMatrix) {
+    }
+}
+
+static bool normalize(std::vector<Dependence::DVEntry> &DV, ScalarEvolution *SE,
```
I am learning this transformation and came across your patch. 

This function is sort of duplicate of FullDependece::normalize() but does not update the distance vector. Do you think we should avoid duplication? and updating distance vector is not necessary?
